runtime/src/bank.rs

```diff
@@ -691,77 +691,106 @@ impl Bank {
             let inflation = self.inflation.read().unwrap();
 
             (*inflation).validator(year) * self.capitalization() as f64 * period
-        };
+        } as u64;
+
+        let vote_balance_and_staked = self.stakes.read().unwrap().vote_balance_and_staked();
 
-        let validator_points = self.stakes.write().unwrap().claim_points();
-        let validator_point_value =
-            self.check_point_value(validator_rewards / validator_points as f64);
+        let validator_point_value = self.pay_validator_rewards(validator_rewards);
+
+        // this sysvar could be retired...
         self.update_sysvar_account(&sysvar::rewards::id(), |account| {
             sysvar::rewards::create_account(
                 self.inherit_sysvar_account_balance(account),
                 validator_point_value,
             )
         });
 
-        let validator_rewards = self.pay_validator_rewards(validator_point_value);
+        let validator_rewards_paid =
+            self.stakes.read().unwrap().vote_balance_and_staked() - vote_balance_and_staked;
+
+        // verify that we didn't pay any more than we expected to
+        assert!(validator_rewards >= validator_rewards_paid);
 
         self.capitalization
-            .fetch_add(validator_rewards as u64, Ordering::Relaxed);
+            .fetch_add(validator_rewards_paid, Ordering::Relaxed);
     }
 
     /// iterate over all stakes, redeem vote credits for each stake we can
     ///   successfully load and parse, return total payout
-    fn pay_validator_rewards(&mut self, point_value: f64) -> u64 {
+    fn pay_validator_rewards(&mut self, rewards: u64) -> f64 {
         let stake_history = self.stakes.read().unwrap().history().clone();
-        let mut validator_rewards = HashMap::new();
 
-        let total_validator_rewards = self
-            .stake_delegations()
+        let stake_delegations = self.stake_delegations();
+
+        // first, count total points
+        let points: u128 = stake_delegations
             .iter()
             .map(|(stake_pubkey, delegation)| {
+                match (
+                    self.get_account(&stake_pubkey),
+                    self.get_account(&delegation.voter_pubkey),
+                ) {
+                    (Some(stake_account), Some(vote_account)) => stake_state::calculate_points(
+                        &stake_account,
+                        &vote_account,
+                        Some(&stake_history),
```

Comment:
It looks like `calculate_points` can be changed to take `&StakeHistory` instead of `Option<&StakeHistory>`, eliminating accidental risk of passing `None` within production codebase.

Reply:
Also, this alleviates code reviewer/reader's cognitive load to some extent. :)

Reply:
lots of tests that don't have a StakeHistory depend on the option